rama-http/src/headers/x_robots_tag_components/robots_tag.rs

```diff
@@ -0,0 +1,214 @@
+            all: false,
+            no_index: false,
+            no_follow: false,
+            none: false,
+            no_snippet: false,
+            index_if_embedded: false,
+            max_snippet: 0,
+            max_image_preview: None,
+            max_video_preview: None,
+            no_translate: false,
+            no_image_index: false,
+            unavailable_after: None,
+            no_ai: false,
+            no_image_ai: false,
+            spc: false,
+            custom_rules: vec![],
+        }
+    }
+
+    pub(super) fn add_custom_rule(&mut self, rule: CustomRule) -> &mut Self {
+        self.custom_rules.push(rule);
+        self
+    }
+
+    pub(super) fn builder() -> Builder {
+        Builder::new()
+    }
+
+    getter_setter!(bot_name, HeaderValueString, optional);
+    getter_setter!(all, bool);
+    getter_setter!(no_index, bool);
+    getter_setter!(no_follow, bool);
+    getter_setter!(none, bool);
+    getter_setter!(no_snippet, bool);
+    getter_setter!(index_if_embedded, bool);
+    getter_setter!(max_snippet, u32);
+    getter_setter!(max_image_preview, MaxImagePreviewSetting, optional);
+    getter_setter!(max_video_preview, u32, optional);
+    getter_setter!(no_translate, bool);
+    getter_setter!(no_image_index, bool);
+    getter_setter!(unavailable_after, ValidDate, optional);
+    getter_setter!(no_ai, bool);
+    getter_setter!(no_image_ai, bool);
+    getter_setter!(spc, bool);
+
+    pub(super) fn is_valid_field_name(field_name: &str) -> bool {
+        field_name.eq_ignore_ascii_case("all")
+            || field_name.eq_ignore_ascii_case("noindex")
+            || field_name.eq_ignore_ascii_case("nofollow")
+            || field_name.eq_ignore_ascii_case("none")
+            || field_name.eq_ignore_ascii_case("nosnippet")
+            || field_name.eq_ignore_ascii_case("indexifembedded")
+            || field_name.eq_ignore_ascii_case("notranslate")
+            || field_name.eq_ignore_ascii_case("noimageindex")
+            || field_name.eq_ignore_ascii_case("noai")
+            || field_name.eq_ignore_ascii_case("noimageai")
+            || field_name.eq_ignore_ascii_case("spc")
+    }
+
+    pub(super) fn from_str(s: &str) -> Result<Option<Self>, OpaqueError> {
+        let mut bot_name = None;
+        let mut fields = s;
+
+        if let Some((bot_name_candidate, rest)) = s.split_once(':') {
+            if !RobotsTag::is_valid_field_name(bot_name_candidate) {
+                bot_name = Some(
+                    HeaderValueString::from_str(bot_name_candidate)
+                        .map_err(OpaqueError::from_std)?,
+                );
+                fields = rest;
+            }
+        }
+
+        let mut builder = RobotsTag::builder().bot_name(bot_name);
+
+        for field in fields.split(',') {
+            match builder.add_field(field) {
+                Ok(_) => {}
+                Err(e) if e.to_string().contains("not a valid robots tag field") => {
```

Comment:
never ever ever string compare errors, use ideally static typed errors for this or in worst case use dynamic checks/casting, but no string checks